src/snabbdom.ts

```diff
@@ -252,45 +275,45 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
 
   function patchVnode(oldVnode: VNode, vnode: VNode, insertedVnodeQueue: VNodeQueue) {
     let i: any, hook: any;
-    if (isDef(i = vnode.data) && isDef(hook = i.hook) && isDef(i = hook.prepatch)) {
+    if (isDef(i = vnode.data)) hook = i.hook;
+    if (isDef(hook) && isDef(i = hook.prepatch)) {
       i(oldVnode, vnode);
     }
     const elm = vnode.elm = (oldVnode.elm as Node);
-    let oldCh = oldVnode.children;
-    let ch = vnode.children;
     if (oldVnode === vnode) return;
     if (vnode.data !== undefined) {
       for (i = 0; i < cbs.update.length; ++i) cbs.update[i](oldVnode, vnode);
-      i = vnode.data.hook;
-      if (isDef(i) && isDef(i = i.update)) i(oldVnode, vnode);
+      if (isDef(hook) && isDef(i = hook.update)) i(oldVnode, vnode);
     }
+    let oldCh = oldVnode.children;
+    let ch = vnode.children;
     if (isUndef(vnode.text)) {
       if (isDef(oldCh) && isDef(ch)) {
-        if (oldCh !== ch) updateChildren(elm, oldCh as Array<VNode>, ch as Array<VNode>, insertedVnodeQueue);
+        if (oldCh !== ch) updateChildren(elm, oldCh as VNodes, ch as VNodes, insertedVnodeQueue);
       } else if (isDef(ch)) {
         if (isDef(oldVnode.text)) api.setTextContent(elm, '');
-        addVnodes(elm, null, ch as Array<VNode>, 0, (ch as Array<VNode>).length - 1, insertedVnodeQueue);
+        addVnodes(elm, null, ch as VNodes, 0, (ch as VNodes).length - 1, insertedVnodeQueue);
       } else if (isDef(oldCh)) {
-        removeVnodes(elm, oldCh as Array<VNode>, 0, (oldCh as Array<VNode>).length - 1);
+        removeVnodes(elm, oldCh as VNodes, 0, (oldCh as VNodes).length - 1);
       } else if (isDef(oldVnode.text)) {
         api.setTextContent(elm, '');
       }
     } else if (oldVnode.text !== vnode.text) {
-      api.setTextContent(elm, vnode.text as string);
+      api.setTextContent(elm, vnode.text);
     }
     if (isDef(hook) && isDef(i = hook.postpatch)) {
       i(oldVnode, vnode);
     }
   }
 
-  return function patch(oldVnode: VNode | Element, vnode: VNode): VNode {
-    let i: number, elm: Node, parent: Node;
-    const insertedVnodeQueue: VNodeQueue = [];
-    for (i = 0; i < cbs.pre.length; ++i) cbs.pre[i]();
-
-    if (!isVnode(oldVnode)) {
-      oldVnode = emptyNodeAt(oldVnode);
+  const patch = function patch(oldVnode, vnode) {
+    if (!isVNode(oldVnode) || !isVNode(vnode)) {
+      throw new TypeError();
```

Comment:
I have 2 issues: 

* `isVnode` returns true against a DOM Element (`document.createElement('div')`) which make sense when you look at the test inside the function...
* Why remove the support of `Element` ? I think it wouldnt cost that much to include the call to `toVnode` inside this lib if the oldVnode is an element instead of introduce a breaking change a make everyone duplicate & include this piece of code

Reply:
yeah, that part was discussed on the initial proposal, snabbdom doesn't know about elements anymore, it knows about vnodes, this can be done in userland code before calling to patch, and we offer a utility function for it.